cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -58,9 +61,48 @@ struct GcsPath {
   }
 };
 
-}  // namespace
+class GcsInputStream : public arrow::io::InputStream {
+ public:
+  explicit GcsInputStream(gcs::ObjectReadStream stream) : stream_(std::move(stream)) {}
 
-namespace gcs = google::cloud::storage;
+  ~GcsInputStream() override = default;
+
+  Status Close() override {
+    stream_.Close();
+    return Status::OK();
+  }
+
+  Result<int64_t> Tell() const override {
+    if (!stream_) {
+      return Status::IOError("invalid stream");
+    }
+    return stream_.tellg();
+  }
+
+  bool closed() const override { return !stream_.IsOpen(); }
+
+  Result<int64_t> Read(int64_t nbytes, void* out) override {
+    stream_.read(static_cast<char*>(out), nbytes);
+    if (!stream_.status().ok()) {
+      return internal::ToArrowStatus(stream_.status());
+    }
+    return stream_.gcount();
+  }
+
+  Result<std::shared_ptr<Buffer>> Read(int64_t nbytes) override {
+    ARROW_ASSIGN_OR_RAISE(auto buffer, arrow::AllocateResizableBuffer(nbytes));
+    stream_.read(reinterpret_cast<char*>(buffer->mutable_data()), nbytes);
+    if (!stream_.status().ok()) {
+      return internal::ToArrowStatus(stream_.status());
+    }
+    return arrow::SliceMutableBufferSafe(std::move(buffer), 0, stream_.gcount());
```

Comment:
Hmm. This is merely slicing the output buffer, but not resizing it. If the actual size is much smaller than `nbytes`, then there is a significant memory wastage.

Reply:
True. On the other hand, a short read from a large buffer could require a lot of copying.  I am not sure which is more likely to be a problem.  I am happy to make a change if you have some data or intuition as to which one is worse.

Reply:
Does GCS return short reads except for the EOF case?

Reply:
For all practical purposes, only the EOF case.  The exception would be some kind of transient error that is stubborn enough to exhaust the retry policy.

Reply:
Ok. In any case, other implementations (such as S3) do resize the buffer, and there doesn't seem to be a reason to act differently here.

Reply:
Created https://issues.apache.org/jira/browse/ARROW-14559 to track that change.